_posts/2019-02-27-vue-with-typescript.md

```diff
@@ -0,0 +1,1331 @@
+			title: 'Look I am blogging!',
+			body: 'Hurray for me, this is my second post!',
+			author: 'Elke',
+			datePosted: new Date(2019, 1, 19),
+		},
+		{
+			title: 'Another one?!',
+			body: 'Another one!',
+			author: 'Elke',
+			datePosted: new Date(2019, 1, 20),
+		},
+	];
+}
+</script>
+
+<style lang="scss">
+#app {
+	font-family: 'Avenir', Helvetica, Arial, sans-serif;
+	-webkit-font-smoothing: antialiased;
+	-moz-osx-font-smoothing: grayscale;
+	text-align: center;
+	color: #2c3e50;
+	margin-top: 60px;
+}
+</style>
+{% endraw %}
+```
+
+As you can see we define a property on the App component that contains our blog posts, we add the components property to the Component decorator and add the BlogPost tag in the template.
+We simply loop over the blog posts with the v-for directive.
+We pass each blog post to the BlogPost component by binding it to the correct data attribute.
+This can be done through `v-bind:post="blogPost"` but we use the shorthand method of `:post="blogPost"`.
+Vue transforms `:post` to `v-bind:post` behind the screens.
+
+Note that we also pass `:key` which we bind to the title of our blog post.
+The reason for this is that it allows Vue to keep track of the state of the list by only looking at the `:key` attribute instead of having to deep compare objects.
+Try to have an unique key.
+A blog post title is a good start but when ordering, updating or other modifying operations you might not have the wanted result if we have blog posts with the same title.
+It's best to use something of type number or string as the key.
+Vue will tell you this in the console of your browser if you would take for example the `datePosted` as your key:
+
+<div style="text-align: center;">
+  <img src="/img/vue-with-typescript/vue-key-typing-error.png" class="image fit">
+</div>
+
+The reason behind this is that Vue relies on the built-in sort and find functionalities of JavaScript.
+For objects, Vue can not do this natively.
+
+When we serve our app with 
+```text
+$ npm run serve
+```
+We see in our browser: 
+
+<div style="text-align: center;">
+  <img src="/img/vue-with-typescript/wordvue-first-impression.png" class="image">
+</div>
+
+Great, you've written your first working component!
+Now it is time to extend it with some functionalities.
+
+## Adding conditional elements to the component
+An important part of a component is to have some dynamic behaviour.
+For example what if we want to show a highlighted blog post? 
+We could create a new component called `HighlightedBlogPost` but we could also extend our existing component.
+
+We can add a new paragraph with a `v-if` statement:
+
+```html
+<p v-if="post.highlighted">This post is highlighted!</p>
+```
+
+The contents of the `v-if` is a TypeScript statement that should return `true` or `false`.
+We extend our `Post` interface to accomodate this:
+
+```typescript
+export interface Post {
+	title: string;
+	body: string;
+	author: string;
+	datePosted: Date;
+	highlighted: boolean;
```

Comment:
Doesn't this need to be optional? Since you are only adding the highlighted property to the highlighted post?